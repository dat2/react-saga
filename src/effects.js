export function render(node) {
  return {
    type: 'RENDER',
    node
  }
}

export function delay(ms) {
  return {
    type: 'DELAY',
    ms
  }
}

export function takeProps() {
  return {
    type: 'TAKE_PROPS'
  }
}

export function call(fn, ...args) {
  return {
    type: 'CALL',
    fn,
    args
  }
}
