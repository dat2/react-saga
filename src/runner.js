import React from 'react'

function run(createGenerator) {
  class Runner extends React.Component {
    constructor(...args) {
      super(...args)
      this.stepGenerator = this.stepGenerator.bind(this)
      this.getNextState = this.getNextState.bind(this)

      this.generator = createGenerator()
      this.state = this.getNextState()
    }

    componentDidMount() {
      this.stepGenerator()
    }

    stepGenerator() {
      this.setState(this.getNextState)
    }

    getNextState(prevState) {
      const { value: effect, done } = this.generator.next()
      if (done) {
        return { done }
      } else if (effect.type === 'RENDER') {
        return {
          done,
          node: effect.node
        }
      } else if (effect.type === 'DELAY') {
        setTimeout(this.stepGenerator, effect.ms)
        return {
          done
        }
      } else {
        return {
          done
        }
      }
    }

    render() {
      return this.state.node
    }
  }

  return Runner
}

export default run
