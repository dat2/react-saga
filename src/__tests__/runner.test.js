import React from 'react'
import { shallow } from 'enzyme'

import runner from '../runner'
import { delay, render } from '../effects'

jest.useFakeTimers()

describe('runner', () => {
  const EmptyRender = runner(function*() {})

  const InitialRender = runner(function*() {
    yield render(<div>initial render</div>)
  })

  const MultipleInitialRenders = runner(function*() {
    yield render(<div>initial render</div>)
    yield render(<div>second initial render</div>)
  })

  it('works for empty sagas', () => {
    const wrapper = shallow(<EmptyRender />)
    expect(wrapper).toMatchSnapshot()
  })

  it('works for the initial render', () => {
    const wrapper = shallow(<InitialRender />)
    expect(wrapper).toMatchSnapshot()
  })

  it('works for multiple initial renders', () => {
    const wrapper = shallow(<MultipleInitialRenders />)
    expect(wrapper).toMatchSnapshot()
  })

  const DelayedInitialRender = runner(function*() {
    yield delay(1000)
    yield render(<div>initial render</div>)
  })

  const DelayedMultipleRenders = runner(function*() {
    yield delay(1000)
    yield render(<div>initial render</div>)
    yield delay(1000)
    yield render(<div>second render</div>)
  })

  it('works when the initial render is delayed', () => {
    const wrapper = shallow(<DelayedInitialRender />)
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1000)
    expect(wrapper).toMatchSnapshot()
  })

  it('works with multiple delayed renders', () => {
    const wrapper = shallow(<DelayedMultipleRenders />)
    jest.runAllTimers()
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })
})
