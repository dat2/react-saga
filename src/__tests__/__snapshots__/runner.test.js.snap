// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runner works for empty sagas 1`] = `""`;

exports[`runner works for multiple initial renders 1`] = `
<div>
  second initial render
</div>
`;

exports[`runner works for the initial render 1`] = `
<div>
  initial render
</div>
`;

exports[`runner works when the initial render is delayed 1`] = `
<div>
  initial render
</div>
`;

exports[`runner works with multiple delayed renders 1`] = `
<div>
  second render
</div>
`;
